fpga_interchange/examples/chipdb.cmake

```diff
@@ -0,0 +1,187 @@
+
+    set(options)
+    set(oneValueArgs device part)
+    set(multiValueArgs)
+
+    cmake_parse_arguments(
+        create_rapidwright_device_db
+        "${options}"
+        "${oneValueArgs}"
+        "${multiValueArgs}"
+        ${ARGN}
+    )
+
+    set(device ${create_rapidwright_device_db_device})
+    set(part ${create_rapidwright_device_db_part})
+    set(rapidwright_device_db ${CMAKE_CURRENT_BINARY_DIR}/${part}.device)
+    add_custom_command(
+        OUTPUT ${rapidwright_device_db}
+        COMMAND
+            RAPIDWRIGHT_PATH=${RAPIDWRIGHT_PATH}
+            ${INVOKE_RAPIDWRIGHT}
+            com.xilinx.rapidwright.interchange.DeviceResourcesExample
+            ${part}
+        DEPENDS
+            ${INVOKE_RAPIDWRIGHT}
+    )
+
+    add_custom_target(rapidwright-${device}-device DEPENDS ${rapidwright_device_db})
+    set_property(TARGET rapidwright-${device}-device PROPERTY LOCATION ${rapidwright_device_db})
+endfunction()
+
+function(create_patched_device_db)
+    # ~~~
+    # create_patched_device_db(
+    #    device <common device>
+    #    patch_name <patch_name>
+    #    patch_path <patch_path>
+    #    patch_format <patch_format>
+    #    patch_data <patch_data>
+    #    input_device <input_device target>
+    # )
+    # ~~~
+    #
+    # Generates a patched device database starting from an input device
+    #
+    # Targets generated:
+    #   - <patch_name>-<device>-device
+
+    set(options)
+    set(oneValueArgs device patch_name patch_path patch_format patch_data input_device)
+    set(multiValueArgs)
+
+    cmake_parse_arguments(
+        create_patched_device_db
+        "${options}"
+        "${oneValueArgs}"
+        "${multiValueArgs}"
+        ${ARGN}
+    )
+
+    set(device ${create_patched_device_db_device})
+    set(patch_name ${create_patched_device_db_patch_name})
+    set(patch_path ${create_patched_device_db_patch_path})
+    set(patch_format ${create_patched_device_db_patch_format})
+    set(patch_data ${create_patched_device_db_patch_data})
+    set(input_device ${create_patched_device_db_input_device})
+
+    get_property(input_device_loc TARGET ${input_device} PROPERTY LOCATION)
+    set(output_device_file ${CMAKE_CURRENT_BINARY_DIR}/${device}_${patch_name}.device)
+    add_custom_command(
+        OUTPUT ${output_device_file}
+        COMMAND
+            ${PYTHON_EXECUTABLE} -mfpga_interchange.patch
+                --schema_dir ${INTERCHANGE_SCHEMA_PATH}
+                --schema device
+                --patch_path ${patch_path}
+                --patch_format ${patch_format}
+                ${input_device_loc}
+                ${patch_data}
+                ${output_device_file}
+        DEPENDS
+            ${input_device}
+            ${input_device_loc}
+    )
+
+    add_custom_target(${patch_name}-${device}-device DEPENDS ${output_device_file})
+    set_property(TARGET ${patch_name}-${device}-device PROPERTY LOCATION ${output_device_file})
+endfunction()
+
+function(generate_chipdb)
```

Comment:
This is specific to 7-series, so name as such?

Reply:
I don't think this function is right.  I think generate chipdb should only be device database -> chipdb.bba

Reply:
Separated in a xc7 specific (which calls RW and the patch funcs) and a generic generate_db which takes as input a device_target